File: src/main/java/top/craft_hello/tpa/utils/VersionUtil.java
package top.craft_hello.tpa.utils;

import org.bukkit.command.CommandSender;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import top.craft_hello.tpa.exceptions.CannotGetTheLatestVersionErrorException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;

import static java.util.Objects.isNull;


/*
* 版本更新检查工具
* 代码来源：EasyLibrary
* 代码原作者：@ZhangYang0204
* 作者Github：https://github.com/ZhangYang0204
* https://github.com/ZhangYang0204/easylibrary (现已不存在)
* Fork：https://github.com/WarSkyGod/easy-gui-shop
* */
public class VersionUtil {
   private static String pluginName;
   private static String pluginVersion;


   public static void init(Plugin plugin) {
      PluginDescriptionFile description = plugin.getDescription();
      pluginName = description.getName();
      pluginVersion = description.getVersion();
   }


   public static @Nullable String readFirstLine(URL url) throws IOException {
      if (isNull(url)) {
         throw new NullPointerException();
      } else {
         InputStream is = url.openStream();
         BufferedReader br = new BufferedReader(new InputStreamReader(is));
         return br.readLine();
      }
   }

   public static void updateCheck(CommandSender sender)  {
      SendMessageUtil.checkUpdate(sender);
      final String latestVersion;
      String latestVersion1;
      try {
         String urlStringPart1 = "https://warskygod.github.io/";
         String urlStringPart2 =  pluginName;
         String urlStringPart3 =  "/latest.html";
         latestVersion1 = readFirstLine(new URL(urlStringPart1 + urlStringPart2 + urlStringPart3));
      } catch (Throwable var5) {
         throw new CannotGetTheLatestVersionErrorException(sender);
      }

      latestVersion = latestVersion1;

      if (versionComparison(pluginVersion, latestVersion)) {
         SendMessageUtil.pluginUpdateMessage(sender, latestVersion);
         return;
      }
      SendMessageUtil.pluginLatestVersion(sender);
   }

   public static String getPluginVersion() {
      return pluginVersion;
   }

   public static int getPluginBigVersion(@NotNull String version) {
      String[] versions = version.split("\\.");
      return Integer.parseInt(versions.length >= 1 ? versions[0] : "0");
   }

   public static int getPluginMiddleVersion(@NotNull String version) {
      String[] versions = version.split("\\.");
      return Integer.parseInt(versions.length >= 2 ? versions[1] : "0");
   }

   public static int getPluginSmallVersion(@NotNull String version) {
      String[] versions = version.split("\\.");
      return Integer.parseInt(versions.length == 3 ? versions[2] : "0");
   }

   public static boolean versionComparison(String version1, String version2){
      int version1Big = getPluginBigVersion(version1);
      int version1Middle = getPluginMiddleVersion(version1);
      int version1Small = getPluginSmallVersion(version1);
      int version2Big = getPluginBigVersion(version2);
      int version2Middle = getPluginMiddleVersion(version2);
      int version2Small = getPluginSmallVersion(version2);
      return isOlderThan(version1Big, version1Middle, version1Small, version2Big, version2Middle, version2Small);
   }

   public static boolean isOlderThan(int currentBig, int currentMiddle, int currentSmall, int big, int middle, int small) {
      if (currentBig > big) {
         return false;
      } else if (currentBig < big) {
         return true;
      } else if (currentMiddle > middle) {
         return false;
      } else if (currentMiddle < middle) {
         return true;
      } else {
         return currentSmall < small;
      }
   }
}

File: src/main/kotlin/top/craft_hello/tpa/utils/VersionUtil.kt
package top.craft_hello.tpa.utils

import cn.handyplus.lib.adapter.HandySchedulerUtil
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import top.craft_hello.tpa.TPA
import top.craft_hello.tpa.objects.ConfigManager
import java.net.HttpURLConnection
import java.net.URI

// 更新检查：直接拉取 GitHub Releases Latest（api.github.com），比较 tag_name 与插件版本
object VersionUtil {
    private const val API_URL = "https://api.github.com/repos/WarSkyGod/TPA/releases/latest"

    private lateinit var plugin: TPA
    private var oldVersion: Boolean = false

    fun init(plugin: TPA) {
        this.plugin = plugin
        oldVersion = versionComparison(plugin.description.version, "4.0")
    }

    fun isOldVersion(): Boolean = oldVersion

    // 比较两段式/三段式版本号：old > new 返回 true
    fun versionComparison(old: String, new: String): Boolean {
        val oldParts = old.substringBefore("-").split(".")
        val newParts = new.substringBefore("-").split(".")
        val size = maxOf(oldParts.size, newParts.size)
        for (i in 0 until size) {
            val o = oldParts.getOrNull(i)?.toIntOrNull() ?: 0
            val n = newParts.getOrNull(i)?.toIntOrNull() ?: 0
            if (o > n) return true
            if (o < n) return false
        }
        return false
    }

    // 启动时后台检查一次（update_check 开启时）
    fun startAsyncUpdateCheck() {
        if (!ConfigManager.config.updateCheck) return
        HandySchedulerUtil.runTaskAsynchronously {
            updateCheck(Bukkit.getConsoleSender())
        }
    }

    // 检查更新并通知 sender（应在异步线程调用）
    fun updateCheck(sender: CommandSender) {
        SendMessageUtil.checkUpdate(sender)
        val latestVersion = fetchLatestVersion() ?: run {
            SendMessageUtil.updateFailedError(sender)
            return
        }
        if (versionComparison(latestVersion, plugin.description.version)) {
            SendMessageUtil.pluginUpdateMessage(sender, latestVersion)
        } else {
            SendMessageUtil.pluginLatestVersion(sender)
        }
    }

    // 拉取 GitHub Latest 的 tag_name；网络异常返回 null
    private fun fetchLatestVersion(): String? {
        var connection: HttpURLConnection? = null
        return try {
            connection = (URI.create(API_URL).toURL().openConnection() as HttpURLConnection).apply {
                connectTimeout = 8000
                readTimeout = 8000
                requestMethod = "GET"
                setRequestProperty("User-Agent", "TPA-Plugin/${plugin.description.version}")
                setRequestProperty("Accept", "application/vnd.github+json")
            }
            if (connection.responseCode != HttpURLConnection.HTTP_OK) return null
            val body = connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            // 不用 gson：1.8 服务器内嵌 gson 2.2.4 没有 JsonParser.parseString（编译期
            // classpath 被 paper-api 携带的新版 gson 调解提升导致编译通过、运行时炸），
            // 只需 tag_name，正则提取即可
            return Regex("\"tag_name\"\\s*:\\s*\"([^\"]+)\"").find(body)
                ?.groupValues?.get(1)?.removePrefix("v")?.removePrefix("V")
        } catch (ignored: Exception) {
            null
        } finally {
            connection?.disconnect()
        }
    }
}
